src/Metadata/CIP25.purs

```diff
@@ -0,0 +1,261 @@
+  :: MintingPolicyHash
+  -> TokenName
+  -> Json
+  -> Either Json.JsonDecodeError CIP25MetadataEntry
+metadataEntryDecodeJson_ policyId assetName =
+  Json.caseJsonObject errExpectedObject $ \obj -> do
+    imageUris <- obj .: "image" >>=
+      decodeNonEmptyStringArray
+    mediaType <- obj .:? "mediaType" >>=
+      pure <<< bindFlipped decodeByteArray
+    description <- obj .:? "description" >>=
+      maybe (pure mempty) decodeStringArray
+    files <- obj .:? "files" >>= \arr ->
+      maybe (pure mempty) (traverse Json.decodeJson) (Json.toArray =<< arr)
+    pure $
+      wrap { policyId, assetName, imageUris, mediaType, description, files }
+
+--------------------------------------------------------------------------------
+-- CIP25Metadata
+--------------------------------------------------------------------------------
+
+newtype CIP25Metadata = CIP25Metadata (Array CIP25MetadataEntry)
+
+derive instance Generic CIP25Metadata _
+derive instance Newtype CIP25Metadata _
+derive instance Eq CIP25Metadata
+
+instance Show CIP25Metadata where
+  show = genericShow
+
+instance ToData CIP25Metadata where
+  toData (CIP25Metadata entries) = unsafePartial $ toData
+    $ Map.singleton (unsafeMkKey nftMetadataLabel)
+    $ Map.fromFoldable
+    $ flip map groups
+    $ \group ->
+        toData (_.policyId $ NonEmpty.head group) /\
+          (Map.fromFoldable <<< flip map group) \entry ->
+            toData (entry.assetName) /\ metadataEntryToData_ (wrap entry)
+    where
+    groups :: Array (NonEmptyArray CIP25MetadataEntryRecord)
+    groups = flip groupBy (unwrap <$> entries) $
+      \a b -> a.policyId == b.policyId
+
+instance FromData CIP25Metadata where
+  fromData meta = unsafePartial $ do
+    entries <- lookupKey nftMetadataLabel meta >>= case _ of
+      Map mp1 -> map concat
+        $ for mp1
+        $ \(policyId /\ assets) ->
+            case assets of
+              Map mp2 ->
+                for mp2 $ \(assetName /\ contents) ->
+                  metadataEntryFromData_ <$> fromData policyId
+                    <*> fromData assetName
+                    <*> pure contents
+              _ -> Nothing
+      _ -> Nothing
+    wrap <$> sequence entries
+
+instance DecodeJson CIP25Metadata where
+  decodeJson =
+    Json.caseJsonObject errExpectedObject $ \obj -> do
+      policies <- obj .: nftMetadataLabel
+      caseJsonObject policies $ \objPolicies ->
+        map (wrap <<< concat)
+          $ for (objToArray objPolicies)
+          $ \(policyId /\ assets) ->
+              caseJsonObject assets $ \objAssets ->
+                for (objToArray objAssets) $ \(assetName /\ contents) -> do
+                  policyId_ <- decodePolicyId policyId
+                  assetName_ <- decodeAssetName assetName
+                  metadataEntryDecodeJson_ policyId_ assetName_ contents
+    where
+    objToArray :: forall a. FO.Object a -> Array (Tuple String a)
+    objToArray = FO.toUnfoldable
+
+    caseJsonObject
```

Comment:
I might recommend changing the name -- it's not exactly name-shadowing since `Data.Argonaut` has been imported qualified, but it might be confusing nonetheless